%YAML:1.0

#Multiple thread support
multiple_thread: 0

# segmmentation
segment_cloud: 1
horizon_scan: 1800
min_cluster_size: 30
min_line_size: 10
segment_valid_point_num: 5
segment_valid_line_num: 3
segment_theta: 1.0 # Lego-loam: 1.0

# laser parameters
idx_ref: 0
n_scans: 16
num_of_laser: 4

cloud_topic:
   - "/top/rslidar_points"
   - "/front/rslidar_points"
   - "/left/rslidar_points"
   - "/right/rslidar_points"

# window sizes
window_size: 6
opt_window_size: 3

# Extrinsic parameter between IMU and Camera.
estimate_extrinsic: 2   # 0  Have an accurate extrinsic parameters. We will trust the following imu^R_cam, imu^T_cam, don't change it.
                        # 1  Have an initial guess about extrinsic parameters. We will optimize around your initial guess.
                        # 2  Have no prior about extrinsic parameters. We will initialize and optimize around them
optimal_extrinsic: 0

# qx qy qz qw px py pz
body_T_laser: !!opencv-matrix
   rows: 4
   cols: 7
   dt: d
   data: [0, 0, 0, 1, 0, 0, 0, 
          0, 0, 0, 1, 0, 0, 0,
          0, 0, 0, 1, 0, 0, 0,
          0, 0, 0, 1, 0, 0, 0] # no prior
   # data: [0, 0, 0, 1, 0, 0, 0, 
   #        -0.0129, 0.0691, 0.0378, 0.9968, -0.1520, -0.9552, 0,
   #        -0.1454, 0.1705, 0.7118, 0.6657, -0.7484, -0.2822, 0,
   #        0.1149, 0.1064, -0.6298, 0.7608, -0.0866, -1.4654, 0] # initialization
   # data: [0, 0, 0, 1, 0, 0, 0, 
   #        -0.00735, 0.05863, 0.02083, 0.99804, 0.53297, 0.04047, -1.06245,
   #        -0.10261, 0.18253, 0.68802, 0.69483, 0.42654, 0.66595, -0.82217,
   #        0.06044, 0.14175, -0.75835, 0.63337, 0.4406, -0.628, -1.0295] # refinement_1
   # data: [0, 0, 0, 1, 0, 0, 0, 
   #        -0.0169, 0.0575, 0.0195, 0.998, 0.5355, 0.0393, -1.131,
   #        -0.1118, 0.1894, 0.6845, 0.6951, 0.5116, 0.6440, -0.904,
   #        0.0745, 0.1312, -0.7449, 0.6496, 0.4406, -0.628, -1.0295] # refinement_2, more accurate
   # data: [0, 0, 0, 1, 0, 0, 0,
   #        0, 0.087150, 0, 0.99679, 0.795, 0, -1.364,
   #        -0.061628, 0.061628, 0.70442, 0.70442, 0.503, 0.6185, -1.0912,
   #        0.061628, 0.061628, -0.70442, 0.70442, 0.503, -0.6185, -1.0912] # CAD

#unsynchronization parameters
estimate_td: 0                      # 0 have an accurate time offset among sensors
                                    # 1 online estimate time offset among sensors

td: !!opencv-matrix                 # initial value of time offset. unit: s. readed image clock + td = real image clock (IMU clock)
   rows: 1
   cols: 4
   dt: d
   data: [0, 0, 0, 0]

distortion: 0
scan_period: 0.1
laser_sync_threshold: 0.07

######################################################## odometry
# optimization 
max_solver_time: 0.02  # max solver itration time (s), to guarantee real time
max_num_iterations: 5   # max solver itrations, to guarantee real time

roi_range: 0.5
distance_sq_threshold: 25
nearby_scan: 2.5

# movement type
planar_movement: 1

# feature corresponding paprameters
min_match_sq_dis: 1.0
min_plane_dis: 0.2

# factor
marginalization_factor: 1
point_plane_factor: 1
point_edge_factor: 1
prior_factor: 0
prior_factor_pos: 0.5
prior_factor_rot: 1

pcl_viewer: 0
pcl_viewer_normal_ratio: 10

# calibration converage parameters
n_cumu_feature: 20
lambda_initial: 100
lambda_thre_calib: 70
n_calib: 15

# good feature
odom_gf_ratio: 1.0

skip_num_odom_pub: 2

######################################################## mapping
map_corner_res: 0.2
map_surf_res: 0.4
map_eig_thre: 100

uct_ext_ratio: 0.5

# uncertainty parameter
# translation, rotation, point (rho, phi, zeta)
uct_ext: !!opencv-matrix
   rows: 4
   cols: 6
   dt: d
   data: [0, 0, 0, 0, 0, 0,
          0.0025, 0.0025, 0.0025, 0.01, 0.01, 0.01,
          0.0025, 0.0025, 0.0025, 0.01, 0.01, 0.01,
          0.0025, 0.0025, 0.0025, 0.01, 0.01, 0.01] 

uct_measurement: !!opencv-matrix
   rows: 1
   cols: 3
   dt: d
   data: [0.0025, 0.0025, 0.0025]

# RV: 20
trace_threshold_before_mapping: 20 # min_d: 20
trace_threshold_after_mapping: 20

